pallene/constant_propagation.lua

```diff
@@ -16,128 +16,180 @@ local function is_constant_value(v)
+
+    for f_id, func in ipairs(module.functions) do
+        -- DFS traversal to find the ir.Cmd.Move instructions which have constant
+        -- values as their src. We can look inside initializers in loops and if-statements because
+        -- the `uninitialized.lua` pass takes care of variables that are used before being initialized.
+        local f_data = assert(data_of_func[f_id])
+
+        for cmd in ir.iter(func.body) do
             local tag = cmd._tag
-            if     tag == 'ir.Cmd.SetGlobal' then
+            if     tag == "ir.Cmd.Move" then
+                local id = cmd.dst
                 if is_constant_value(cmd.src) then
-                    constant_initializer[cmd.global_id] = cmd.src
+                    f_data.locvar_constant_init[id] = cmd.src
                 end
-            elseif tag == 'ir.Cmd.Seq' then
-                for i = #cmd.cmds, 1, -1 do
-                    table.insert(stack, cmd.cmds[i])
+
+            elseif tag == "ir.Cmd.SetUpvalues" then
+                for u_id, value in ipairs(cmd.srcs) do
+                    local next_f = data_of_func[cmd.f_id]
+                    if value._tag == "ir.Value.LocalVar" then
+                        local const_init = f_data.locvar_constant_init[value.id]
+                        next_f.constant_val_of_upvalue[u_id] = const_init
+
+                    elseif value._tag == "ir.Value.Upvalue" then
+                        local const_init = f_data.constant_val_of_upvalue[value.id]
+                        next_f.constant_val_of_upvalue[u_id] = const_init
+                    end
                 end
-            else
-                -- skip
+
             end
         end
+
+        for loc_id = 1, #func.typ.arg_types do
+            f_data.locvar_constant_init[loc_id] = false
+        end
     end
 
-    -- 2) Find what toplevel variables are never re-initialized or never used
+    -- 2) Find which local variables are never re-initialized.
 
-    local n_reads  = {} -- { g_id => int }
-    local n_writes = {} -- { g_id => int }
-    for i = 1, n_globals do
-        n_reads[i]  = 0
-        n_writes[i] = 0
-    end
+    for f_id, func in ipairs(module.functions) do
+        local f_data   = assert(data_of_func[f_id])
+        local n_writes = f_data.n_writes_of_locvar
 
-    for _, func in ipairs(module.functions) do
         for cmd in ir.iter(func.body) do
             local tag = cmd._tag
-            if     tag == "ir.Cmd.GetGlobal" then
-                local id = cmd.global_id
-                n_reads[id] = n_reads[id] + 1
-            elseif tag == "ir.Cmd.SetGlobal" then
-                local id = cmd.global_id
-                n_writes[id] = n_writes[id] + 1
+            if tag == "ir.Cmd.SetUpvalues" then
+                for u_id, value in ipairs(cmd.srcs) do
+                    local next_f = assert(data_of_func[cmd.f_id])
+
+                    if value._tag == "ir.Value.LocalVar" and n_writes[value.id] ~= 1 then
```

Comment:
since x1 is being reassigned to, wouldn't that make it a mutable upvalue, and therefore not a constant (since the assignment conversion pass turns it into a record). I think our assumption was that all upvalues that are initialized to a constant in the IR are guaranteed to be `<const>` because assignment_conversion pass would box them otherwise.

Reply:
I think the problem is that we're computing the `n_writes` in the same loop that we are computing the `constant_val_of_upvalue`. When we  process the  `x2.upvalues <- x1`, we don't know the total `n_writes` for the `x1` yet.

Reply:
Could you perhaps give a Pallene code snippet as example?
I'm wondering if generating IR like above is even possible. Because if the Pallene source is something like this:

```lua
local x = 10
if x == 10 then
	local function f(): integer return x end
	x = 20
end
```

Then the `x` would become a record type, and so the n_writes wouldn't really matter since constant_val_of_upvalue would be false anyway. This makes me also realize that we are relying on assumptions across multiple modules that are not properly documented anywhere. I wonder what the appropriate place for that would be? Since `manual.md` is meant for programmers who will use Pallene and not developers who will look into it's source.

Reply:
You might be right, this might be another case where it is allowed by the IR but never happens in the IR that we generate.

> This makes me also realize that we are relying on assumptions across multiple modules that are not properly documented anywhere. I wonder what the appropriate place for that would be?

We don't necessarily want to codify these assumptions. While it might be true that the assumptions hold for the output of the `to_ir`, that may not necessarily be the case if we add more optimizations between the `to_ir` and the `constant_propagation`.

I think the ideal thing to do would be to make the optimization more general (accepting all representable IR programs)  or to modify the IR to make the undesirable programs impossible to represent.

A less ideal solution, which is less work,  would be to add some assertions to the optimization pass. That way, if the assumption gets broken later, at least we crash the compiler instead of producing the wrong output.

Reply:
One possibility is to move forward with this assumption by adding some `assert`s for now, since that makes a lot of things easier. We can then open an issue to make it a point that the optimizations don't work on all possible valid IR, just IR that is generated by the `to_ir` pass and verified by the `uninitialized.lua` pass.

Reply:
Agreed. But I wouldn't give a high priority for fixing that issue. There's a decent chance that at least some of the optimizations will have to be re-implemented in the future.